library/src/main/java/SpecTreeToYamlStringConverter.java

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Provides the ability to serialize Spec Trees represented as LinkedHashMaps into YAML strings */
+public class SpecTreeToYamlStringConverter {
+
+  int indent;
+
+  public SpecTreeToYamlStringConverter() {
+    this.indent = 2;
+  }
+
+  /**
+   * Serializes a spec tree represented as a {@code LinkedHashMap} into a YAML string.
+   *
+   * @param yamlMap a spec tree which is a LinkedHashMap with String keys and Object values
+   * @return a YAML string which represents the serialization of {@code yamlMap} as a YAML string
+   */
+  public String convertSpecTreeToYamlString(LinkedHashMap<String, Object> yamlMap)
+      throws UnexpectedTypeException {
+    return convertSpecTreeToYamlString(yamlMap, 0, false);
+  }
+
+  /**
+   * Helper function for Serializing a spec tree represented as a {@code LinkedHashMap} to a YAML
+   * string.
+   *
+   * @param yamlMap a spec tree which is a LinkedHashMap with String keys and Object values
+   * @param level the current traversal level, which influences how many spaces to print out
+   * @param firstListElement a boolean which indicates whether or not current element in the loop is
+   *     the first element of a list, which needs to be processed in a special way
+   * @return a YAML string which represents the serialization of {@code yamlMap} as a YAML string
+   */
+  private String convertSpecTreeToYamlString(
+      LinkedHashMap<String, Object> yamlMap, int level, boolean firstListElement)
+      throws UnexpectedTypeException {
+    StringBuilder str = new StringBuilder();
+    for (Map.Entry<String, Object> entry : yamlMap.entrySet()) {
+      String key = entry.getKey();
+      Object value = entry.getValue();
+
+      firstListElement = handleFirstListElementSpacing(level, firstListElement, str);
+
+      handleKey(str, key);
+
+      if (TypeChecker.isObjectMap(value)) {
```

Comment:
Maybe add a comment for each about what is an object map object list, or object primative, and what could be considered an "Exception" type with adding an example in header or inline.   This will help readers of the open source repo get an idea of "what" you are processing as input on these "is" statements.

Reply:
@ApigeeCorky thanks for the feedback!! In general, would it be best practice to do that here in the serializer class, or within the "TypeChecker" class itself?